Add video panning

moex3 opened the pull request:
Add the ability to move the preview video around, and add a menu item to reset the panning.

It's a little bit hacky, as i got this to work by trial and error.

I tested it on windows with the wangqr fork (as i was able to build that), and on linux with xorg. Both works fine, however on wayland the zoom anchor seems to be in the bottom left corner instead of the upper left one and that breaks the mouse position. But this requires more testing.

It can be used by dragging, with the middle mouse button.
Example ![video](https://user-images.githubusercontent.com/46636583/107129957-9441fd80-68c9-11eb-864a-b0682974da5c.mp4)

closes #187, closes #107

moex3 committed:
Add video panning

moex3 committed:
Add menu item for resetting the video pan

moex3 committed:
Change tab to spaces

moex3 committed:
Fix mouse position bug when videoAr > displayAr

Crosshair drawing is broken when videoAr != displayAr

moex3 committed:
Fix crosshair drawing

I'm not sure about that scale_factor, i don't know how to test that

moex3 committed:
Change pan values with the zoom level changes

moex3 committed:
Respect min. zoom value when changing pan coords

megax commented:
I solved a couple of years ago. I think it's more perfect that way.

https://drive.google.com/file/d/150QXmgTLgTmVBdu0Zbo_3CLPttpbwGMz/view?usp=sharing

moex3 commented:
@megax Yeah that looks good, i was also considering stopping the zoom from changing the preview window size, but haven't got around to it yet (and i don't feel like i'm being pressured, looking at the state of the project).

I tried searching for this feature a while ago, found nothing just github issues with 0 progress on them, so I considered it non-existent. Is it Haru-Subs private :D ? If not, create a PR please.

megax commented:
@moex3 Yeah sorry, the code is private. I exclusively developed it for our group in 2018 and our group and it stays exclusive for the time being. I just sent the video, so you can take a look at what it "should" look like. Our Aegi contains quite a few more features besides this, so it's not a little project. These mechanics took a lot of time to figure out. Sorry for getting you excited, but I really can't publicize it, since it belongs to our team(now Guddo-FanSub).

moex3 committed:
Make video zoom different from preview window zoom

moex3 committed:
Make zooming center around the cursor position

moex3 committed:
Make zooming feel more linear

with a "good enough" implementation

moex3 committed:
Fix a zooming bug on linux

The preview window would change in width, when the video was being zoomed.
Possibly, because on windows, zooming only calls UpdateSize once, but on
linux, it gets called multiple, like 8 times.
Rounding fixed it.

moex3 commented:
Some updates!

Now the video can be zoomed without affecting the displaying window size. Scrollwheel zooms the video, and ctrl+scrollwheel zooms the entire preview window. Zooming centers around the cursor. Zooming now feels more linear, but don't zoom too much, it gets weird. Also, this implementation of it is not perfect, as after a zoom-in, a zoom-out does not result in the original video size. Zoom-zoom.

Preview:

[prev](https://user-images.githubusercontent.com/46636583/115474166-27d66980-a23d-11eb-8be5-15eabb1d8e71.mp4)

megax commented:
@moex3 
Video window default (my code sample) ~ (ctrl+0)
Set:

```
void VideoDisplay::OnKeyDown(wxKeyEvent &event) {
	hotkey::check("Video", con, event);

	// Zoom 1.0
	if(event.ControlDown() && (event.GetKeyCode() == '0' || event.GetKeyCode() == WXK_NUMPAD0)) {
		auto provider = con->project->VideoProvider();
		if (provider && IsShownOnScreen()) {
			isresetzoom = true;
			ResetZoom(); // set zoom value 1.0 etc
			last_videoSize.Set(provider->GetWidth(), provider->GetHeight());
			UpdateSize();
		}
	}
}
```

moex3 committed:
Add reset hotkey as Ctrl-R

I'm not entirely sure, that this is how it's supposed to be added.
This requires editing the already existing hotkey.json file (shows path
in log, if it exists).
Should this also be added to hotkey.cpp?

seiya-eyth commented:
Hi! @moex3 ,thanks for this beautiful useful feature, but i don't know how to install it in my aegisub, can you help me please.

moex3 commented:
Hi @seiya-eyth!

You can use [AegisubDC](https://github.com/Ristellise/AegisubDC/releases), which is a fork of aegisub that has this PR merged.

moex3 committed:
Revert "Add reset hotkey as Ctrl-R" / Not needed

This reverts commit 3d525d9938790211b02a31a16d716498957891fc.
The default hotkey is not required. People can just add a custom one
from the hotkey settings if they want this.

seiya-eyth commented:
@moex3 it worked and i can zoom now. But the json and cpp files, where i have to paste them ?

moex3 commented:
@seiya-eyth Those files are only needed if you want to build aegisub from source code. AegisubDC provides already built exe files, so if you downloaded that, and it works, then it's fine. You have nothing else to do.

The json file was used to set a hotkey to reset the video zoom. You can set a custom one for it in the settings if you want that.
